Solutions/Homeworks/homework4.lean

```diff
@@ -0,0 +1,201 @@
+  ext x -- The expression `x ∈ f ⁻¹' p` reduces to `f x ∈ p`
+  rfl
+
+-- If `s` is a set of elements of type `α`, the library also defines `image f s`, written `f '' s`,
+-- to be `{ y | ∃ x, x ∈ s ∧ f x = y }`. So a hypothesis `y ∈ f '' s` decomposes to a triple
+-- `⟨x, xs, xeq⟩` with `x : α` satisfying the hypotheses `xs : x ∈ s` and `xeq : f x = y`.
+-- The `rfl` tag in the `rintro` tactic was made precisely for this sort of situation.
+
+example : f '' (s ∪ t) = f '' s ∪ f '' t := by
+  ext y; constructor
+  · rintro ⟨x, xs | xt, rfl⟩ -- Notice how `rfl` automatically rewrites `f x = y`
+    · left
+      use x, xs
+    right
+    use x, xt
+  rintro (⟨x, xs, rfl⟩ | ⟨x, xt, rfl⟩)
+  · use x, Or.inl xs
+  use x, Or.inr xt
+
+
+example : s ⊆ f ⁻¹' (f '' s) := by
+  intro x xs
+  -- The goal decomposes to
+  use x, xs
+
+-- 1
+example : f '' s ⊆ v ↔ s ⊆ f ⁻¹' v := by
+  constructor
+  . intro h x xs
+    exact h ⟨x, xs, rfl⟩
+  . rintro h x ⟨y, ys, rfl⟩
+    exact h ys
+
+-- 2
+example (h : Injective f) : f ⁻¹' (f '' s) ⊆ s := by
+  rintro x ⟨y, ⟨hy, hxy⟩⟩
+  rw [← h hxy]
+  exact hy
+
+-- 3
+example : f '' (f ⁻¹' u) ⊆ u := by
+  rintro y ⟨x, h, rfl⟩
+  exact h
+
+-- 4
+example : f ⁻¹' u \ f ⁻¹' v ⊆ f ⁻¹' (u \ v) := by
+  rintro x ⟨h₁, h₂⟩
+  exact ⟨h₁, h₂⟩
+
+-- 5
+example : f '' s ∩ v = f '' (s ∩ f ⁻¹' v) := by
+  ext y; constructor
+  · rintro ⟨⟨x, ⟨hxs, rfl⟩⟩, hxy⟩
+    use x, ⟨hxs, hxy⟩
+  . rintro ⟨x, ⟨xs, h⟩, rfl⟩
+    use ?_, h
+    use x, xs
+
+-- 6
+example : s ∩ f ⁻¹' u ⊆ f ⁻¹' (f '' s ∩ u) := by
+  rintro x ⟨hx, hx2⟩
+  use ?_, hx2
+  use x, hx
+
+-- 7
+example : s ∪ f ⁻¹' u ⊆ f ⁻¹' (f '' s ∪ u) := by
+  rintro x (h | h)
+  · left
+    use x, h
+  . right
+    exact h
+
+
+end
+
+
+section
+
+-- Let us consider functions `ℝ → ℝ`
+
+open Set Real
+
+-- `InjOn f s` is a predicate that means the function `f` is injective on `s`
+
+#print InjOn
+
+def PosReals : Set ℝ := { x : ℝ | 0 < x }
+
+example : InjOn log PosReals := by
+  intro x xpos y ypos
+  intro e
+  -- `log x = log y`
+  calc
+    x = exp (log x) := by rw [exp_log xpos]
+    _ = exp (log y) := by rw [e]
+    _ = y := by rw [exp_log ypos]
+
+-- `range f` is defined as the set `{ x : ℝ | ∃ y : ℝ, f y = x }`
+
+#print range
+
+example : range exp = PosReals := by
+  ext y; constructor
+  · rintro ⟨x, rfl⟩
+    apply exp_pos
+  intro ypos
+  use log y
+  rw [exp_log ypos]
+
+-- You may need the following theorems:
+
+#check sqrt_nonneg
+#check pow_nonneg
+
+def NonNegReals : Set ℝ := { x : ℝ | 0 ≤ x }
+
+-- 8
+example : InjOn sqrt NonNegReals := by
+  intro x xpos y ypos hxy
+  rw [← sq_sqrt xpos, hxy, sq_sqrt ypos]
+
+-- 9
+example : InjOn (fun x ↦ x ^ 2) NonNegReals := by
+  intro x xpos y ypos hxy
+  dsimp at *
+  rw [← sqrt_sq xpos, hxy, sqrt_sq ypos]
+
+-- 10
+example : sqrt '' NonNegReals = NonNegReals := by
+  ext y; constructor
+  · rintro ⟨x, _, rfl⟩
+    exact sqrt_nonneg _
+  intro ypos
+  use y ^ 2
+  rw [sqrt_sq ypos]
+  refine ⟨sq_nonneg y, rfl⟩
```

Comment:
I don't think we've mentioned `refine` either. We could decide to introduce it here, or an `exact` would work too.